Improve appearance and readability

NunoAlexandre opened the pull request:
## Motive

I work with haddock and hackage almost daily, both at work and at home, and the current theme is unfriendly. I find it very hard to read content with such a wide `#content`, and spacing is not great either. 

Furthermore, I'd say many Haskell developers would appreciate the presense of the latest Haskell's logo (purple) colors in these docs. 

## Changes

- use latest Haskell's logo colors
- decrease #content width to improve readability
- use, custom nicer font: [Merriweather Sans](https://fonts.google.com/specimen/Merriweather+Sans)
- improve sizes and distances

## Preview

**EDIT 04-02-2018**: See https://github.com/haskell/haddock/pull/721#issuecomment-362934997

<br>
Using: https://hackage.haskell.org/package/lens-tutorial-1.0.3/docs/Control-Lens-Tutorial.html

**Before:**

<img width="1280" alt="before" src="https://user-images.githubusercontent.com/7075260/34361086-6befd534-ea67-11e7-887c-02a8f70272a1.png">


**After:**
<img width="1280" alt="after" src="https://user-images.githubusercontent.com/7075260/34361092-78decd2c-ea67-11e7-93ec-65346b10b792.png">

<img width="337" alt="haddock-after-tablet-standing" src="https://user-images.githubusercontent.com/7075260/34388512-4bf1db88-eb34-11e7-9416-cbd156a0d693.png">

<img width="597" alt="haddock-after-tablet-laying" src="https://user-images.githubusercontent.com/7075260/34388521-52fa1300-eb34-11e7-9dc4-075971ac9ff5.png">

<img width="256" alt="haddock-after-iphone7" src="https://user-images.githubusercontent.com/7075260/34388530-634dcfd0-eb34-11e7-9c3b-e86733df5300.png">

NunoAlexandre commented:
@phadej 

A screenshot of https://hackage.haskell.org/package/lens-4.16/docs/Control-Lens-Operators.html with these changes:
![screenshot from 2018-01-28 21-42-23](https://user-images.githubusercontent.com/7075260/35486868-97e02a14-0474-11e8-9813-2bce6f173078.png)

Some bits are not 100%, I noticed. I hope to work on it tomorrow.

phadej commented:
@NunoAlexandre thanks for the screenshot. It's indeed obvious that some tuning is needed.

phadej commented:
Also we need to check that
- Mathematics/LaTeX (I don't remember whether there are packages on Hackage, does @hvr or @alexbiehl remember) and
- Tables (there aren't yet packages on Hackage using them, but haddock's `master` supports)

are styled nicely. But that won't be a blocker, rather a thing to keep in mind.

---

- I like nice fonts, but I don't like webfonts. caniuse says we can use them https://caniuse.com/#feat=fontface, yet there should be **offline** fallback which is nice enough.

also Info:

```
% curl -sL 'https://fonts.gstatic.com/s/merriweathersans/v9/nAqt4hiqwq3tzCecpgPmVZQnJ6jNVR5ny37t9nSm1vT3rGVtsTkPsbDajuO5ueQw.woff2' | wc
     48     307   15808
```

and there are five (5) variants in the stylesheet.

I don't know whether Hackage would like to host (and whether it can, due font licenses) those on their own CDN, but I don't see bandwidth would be a problem (stylesheet itself is 11k, so styling grows from 11k -> ~70k - not something to worry about).

cocreature commented:
> I don't remember whether there are packages on Hackage, does @hvr or @alexbiehl remember

https://hackage.haskell.org/package/kalman-1.0.0.2/docs/Numeric-Kalman.html uses this quite heavily.

NunoAlexandre commented:
I am doing some further adjustments, namely darkening the color of the text and making it slighter bigger, changing the color of the links to shades of purple instead of orange, etc. I will push those changes tomorrow, but in the meantime here follows a screenshot of the `Control Lenses Operators` page:
 
![haddock-lenses-operators](https://user-images.githubusercontent.com/7075260/35535666-9452e7d0-0544-11e8-94c4-ad3e1fca3ebf.png)

phadej commented:
I can barely see the code blocks background color. Should I calibrate my display?

phadej commented:
@bgamari is https://ghc.haskell.org/trac/ghc/wiki/Commentary/Libraries/VersionHistory, especially `time` will stay at `1.8.0.2` for GHC-8.4?

phadej commented:
Also, @hvr what;'s about `parsec-3.1.12`?

elben commented:
@NunoAlexandre this is in reply to your comments on the [other PR](https://github.com/haskell/hackage-server/pull/648#issuecomment-362404128). Putting it here because my concerns were largely with the Haddock pages.

> I personally cannot recall a case where density was more important than readability, for I would not prefer to see more content at once and have a much harder time reading any of it. I might be wrong here, I am not dismissing your observation.
>
> Do you have an example of a material which is better read with a dense design than with a good negative space that you can show?

I think the current Haddock pages, [Hacker News], [nytimes.com](https://www.nytimes.com) and is a good example where high-density design works. When I'm in Haddock, my brain is usually in "search" mode. I'm looking for some function I need, or trying to get a sense of what is available. In the same way when I'm scrolling HN or nytimes, I'm searching for something to read.

A good Haddock example for this use-case is [Parsec's combinators](https://hackage.haskell.org/package/parsec-3.1.11/docs/Text-Parsec.html#v:-60--124--62-). Can you show an example of what that section would look like with your changes?

> I am not with you on this one. I have minor studies in typography and (negative) space is fundamental, and a lot of the things you are not into or don't understand have been thoroughly studied and tested. You are obviously free to have your preference, but I am afraid your are not on the winning side of the discussion.
> If your point is that I am using too much negative space now, I am open to improve it. But I'd rather have a first version deployed and see what the masses have to say and tweak accordingly, as now the majority seems to like it as I am proposing (which obviously doesn't mean it's right), and democracy rules.

I agree that negative space is important. I also agree that deploy-now-fix-later is a valid approach, as long as we're open and have capacity to improving it later.

I will point out that your [mobile example](https://user-images.githubusercontent.com/7075260/34388530-634dcfd0-eb34-11e7-9c3b-e86733df5300.png) shows where there is too much negative space. The longest line has 42 characters (may be off by one or two), whereas from my understanding good practice is somewhere between 2-3 runs of the full alphabets (26*2). The padding for the body text seems like wasted space to me, and hurts readability flow.

> I agree there is an exageration of low-contrast overall on the web, but I don't see how that is the case here? I think an example of it is, actually, Stackage (which I prefer to look at than Hackage, FWIW), but - again - I can't see how that is a thing on this PR.

I don't have a solid argument or evidence other than I have to squint more with the new contrast, in comparison to the old one.

I wasn't clear before, but overall I think this is a great step forward!

NunoAlexandre commented:
### Update

I have made some adjustments in order to make this stylesheet consistent with the Hackage one.

Below you can see screenshots of all pages you have asked to see. Note as well the redesigned menu for smaller screens. 


<details>

<summary> Screenshots</summary>

**[Control.Lens.Operator](https://hackage.haskell.org/package/lens-4.16/docs/Control-Lens-Operators.html)**
![lenses-folds-traversals-smaller](https://user-images.githubusercontent.com/7075260/35771323-3293c772-092b-11e8-8216-cc5380a5885c.png)

**[Numeric Kalman](https://hackage.haskell.org/package/kalman-1.0.0.2/docs/Numeric-Kalman.html)**:
![kalman](https://user-images.githubusercontent.com/7075260/35771333-5d2c1584-092b-11e8-8307-5aa126870fb1.png)

**[Control.Lens.Cons](https://hackage.haskell.org/package/lens-4.16/docs/Control-Lens-Cons.html)**
![lens-cons](https://user-images.githubusercontent.com/7075260/35771342-7f7d6660-092b-11e8-95d1-fd876481d219.png)

**[Control.Lens.Tutorial](https://hackage.haskell.org/package/lens-tutorial-1.0.3/docs/Control-Lens-Tutorial.html)**
![lens-tutorial](https://user-images.githubusercontent.com/7075260/35771347-957368fc-092b-11e8-8577-9ee01be9fad1.png)


<hr>

@phadej, about your note on webfonts: I get that, but it is well-supported as you noted and I have already reduced the number of variants to only 3. The increase in style assets is minimal indeed.

NunoAlexandre commented:
# Preview latest suggestions

Instead of posting screenshots, I have hosted the pages you guys have shown interest in seeing the latest suggestions here proposed. 

You can find a list of links to them following the link below: 
https://nunoalexandre.com/2018/02/04/redesigning-haskell-docs#preview

This allows anyone to preview and try it out! I will keep on updating it as we ago so it will always be updated.

gbaz commented:
The live previews are very useful. I think there is a problem with the "synopsis" "contents" listings in the haddocks. The contents are sort of as before, but just look much worse in this new setting, as though they're not fully integrated with the design. Also with everything larger and with more whitespace they just displace more. 

![image](https://user-images.githubusercontent.com/1125849/35783964-1fb1663c-09df-11e8-9af5-746b6c93c486.png)

"synopsis" is just unstyled (rather than being in a nice tab) when unexpanded, and when expanded it also lacks some proper styling:

![image](https://user-images.githubusercontent.com/1125849/35783977-593d091a-09df-11e8-9b61-4c0a87e9e668.png)

mightybyte commented:
It makes me really happy to see this.  We could definitely use a more modern refresh to the hackage / haddock design.  Thank you @NunoAlexandre for your work and raising the conversation here.

At the moment I think the information density argument wins it for me.  When I'm looking at haddock pages, I am almost always scanning and want to see more things at once rather than read the page sequentially.  Also, type signatures are the most important part of the content.  I want to see more type signatures on a page along with maybe a slug of the first paragraph of each function's description.

So while the designs you have so far definitely look better, I think they would make haddock docs less usable for me than they are right now.

gbaz commented:
Personally if I want to fit more in a page, I don't mind hitting cmd-minus a few times to zoom out. The obstacle here seems to be that there's still an undue amount of space between code snippets like in the following:

![image](https://user-images.githubusercontent.com/1125849/35785175-840c4bc6-09ec-11e8-9e69-8da114b7dc9b.png)

That sort of whitespace doesn't get affected enough by the zoom feature of browsers, so it would be better to just cut it down, imho.

Outside of that, I don't mind the larger font size (since, as I mentioned, its easy to reduce).  But I also wouldn't mind a slightly smaller one :-)

I _do_ think the wider margins are a good thing here, as in the haddocks as they stand now, we typically don't have more "information density" (since there's only so much code on each line) but rather just big empty bands of grey that make the stuff flush-left and flush-right much harder to visually associate.

gbaz commented:
I'd also like to see how this handles a degenerate case like this (see the comments pushed all the way to the right by the typeclass instances): https://hackage.haskell.org/package/servant-server-0.12/docs/Servant-Server.html#t:HasServer

(relevant ticket -- https://github.com/haskell/haddock/issues/300)

Not to keep inflating scope, but this would be really nice to tackle if its not too invasive to fixup...

NunoAlexandre commented:
@gbaz 

> The contents are sort of as before, but just look much worse in this new setting, as though they're not fully integrated with the design. Also with everything larger and with more whitespace they just displace more.

I am not sure I understand your point here. Are you just suggesting to adjust them to the new design or something deeper? I don't understand how it "looks much worse in this new setting" left alone, thus the question.

> "synopsis" is just unstyled (rather than being in a nice tab) when unexpanded, and when expanded it also lacks some proper styling:

Indeed, I have to fix this. Thanks!

 > I'd also like to see how this handles a degenerate case like this

As soon as I have time I will add it, but it won't be on the top of my list.

Thanks again!

NunoAlexandre commented:
@mightybyte 

> It makes me really happy to see this. We could definitely use a more modern refresh to the hackage / haddock design. Thank you @NunoAlexandre for your work and raising the conversation here.

Thank you very much for your words!

> At the moment I think the information density argument wins it for me. When I'm looking at haddock pages, I am almost always scanning and want to see more things at once rather than read the page sequentially. Also, type signatures are the most important part of the content. I want to see more type signatures on a page along with maybe a slug of the first paragraph of each function's description.

I am sensitive to this argument, but note that  - as in everything - there is a trade-off being made. By increasing information density we lose readability. And the information that these packages display is not exactly mere superficial text but content that pushes your cognitive load forward, therefore improving readability balances this load. The `synopsis` (to be fixed) is there exactly for this purpose: for one to get a complete glance of what functions are documented in the current page and to get to them in a click.

I will definitely polish corners here and there to optimize as much as possible both worlds.

Thanks for your feedback.

NunoAlexandre commented:
@gbaz 

> The obstacle here seems to be that there's still an undue amount of space between code snippets like in the following:

I agree, my bad here. I have fixed this already. Thanks for pointing it out!

> I do think the wider margins are a good thing here, as in the haddocks as they stand now, we typically don't have more "information density" (since there's only so much code on each line) but rather just big empty bands of grey that make the stuff flush-left and flush-right much harder to visually associate.

I make yours my words.

Thanks

gbaz commented:
> I am not sure I understand your point here. Are you just suggesting to adjust them to the new design or something deeper? I don't understand how it "looks much worse in this new setting" left alone, thus the question.

I think it just feels larger and therefore that it displaces more. I'm not quite sure how to fix this -- perhaps it can be left for now. Or maybe the css for the functions can be adjusted so they don't run straight-into the contents, but leave a margin, as the css for the code-blocks does.

Hrothen commented:
Would it be possible to implement these changes as a flag rather than replacing the current theme? I'd like to be able to retain the current style in docs I build for myself.

NunoAlexandre commented:
I have increased the content width a little for all screen sizes and reworked the synopsis. The latter is not great yet but hopefully better and more functional for all use cases.

lthms commented:
@NunoAlexandre you are probably aware of it, but if the synopsis is longer than the screen, you can’t see its end. That being said, I really like what you are doing here and look forward to see it merged and used upstream!

NunoAlexandre commented:
Thanks, @lthms, I am not able to fix it now but it's now on my list :+1:

cloudhead commented:
Beautiful, thanks for this! I hope we can make it work - I do agree with most comments around space utilization - people use haddock as a tool, so the right information density is important.

cloudhead commented:
A few observations on your latest preview:

1. The bolded bits in the code/repl segments with `>>> ...` draw too much attention compared to the rest. I don't think they need to be bolded.
2. The synopsis doesn't scroll.
3. Margin between paragraphs could be smaller to save v-space. Currently looks a bit odd with the one sentence paragraphs which are common.
4. The text on the topmost menubar is not v-aligned, not sure if it's on purpose but it's distracting.
5. The top/bottom padding inside code-blocks could be lesser to save v-space, they are quite big.

igrep commented:
I like most of the design but have one concern: we may lose the way without a link to the index page (top page) on the header.
But unfortunately, it's not achieved by the current haddock/hackage.
Maybe it's not the problem of your design but of the technical design of haddock/hackage themselves...
So I'll file a separate issue if needed.

(I wonder why I've never cared... 😅 ))

gbaz commented:
@igrep i agree that would be a separate issue, but if i understand what you're asking for, that's provided by the "contents" link already?

igrep commented:
@gbaz Oops! Yeah, I missed the flow "Module page -> Contents -> Search box".
That might be sufficient.

adius commented:
@NunoAlexandre Why is the header not also limited to 60 em width? Normally my mouse will be in the middle of the screen and this means I've to move it all the way to the right corner to click contents ...
I think it would look better and make it more usable if the the width of the body and the header were the same ...

adius commented:
Second question: What about syntax highlighting? Did you fix this as well?
I developed a theme for my company:

<img width="1280" alt="2018-02-11_syntax_highligthing_browser" src="https://user-images.githubusercontent.com/969983/36068309-3ca053bc-0eca-11e8-84c9-4de7df80d7ba.png">

It's written in Clay and could be easily adopted to fit your color scheme.
Let me know if this could help you...

NunoAlexandre commented:
### Update
I have addressed the many comments about information density by reducing the font-size a little and improving some spacing between elements.

NunoAlexandre commented:
> @NunoAlexandre Why is the header not also limited to 60 em width? Normally my mouse will be in the middle of the screen and this means I've to move it all the way to the right corner to click contents ...
I think it would look better and make it more usable if the the width of the body and the header were the same ...

@adius, I tried that initially, but since the package name goes on the header, it left very little space. which caused the title to be broken into multiple leans very often.

I highly agree with the syntax highlighting but let's do that in a second PR :+1:

alexbiehl commented:
Alright! Hackage has been released with your new theme. Could you rebase your work once more?

gbaz commented:
note: a few of the changes to `hackage.css` from this patch probably should be ported over: https://github.com/haskell/hackage-server/commit/740ceeac25666123c030fd22b8b8c3b56cdec6e6#diff-066aa644bc1eb7b490f78fd5dc570b79

NunoAlexandre commented:
@alexbiehl, I have rebased from `master` and I have ported over some of the changes made recently to hackage, namely link colors and menu behavior. I've fixed a responsive issue as well.

Let me know if you insist on pointing this branch onto `ghc-8.4` and rebasing it from there.

alexbiehl commented:
Thanks @NunoAlexandre! It's alright, I will pick the changes into ghc-8.4 branch once this is merged.

NunoAlexandre commented:
Cool, @alexbiehl! Who can push the button? Could you, or does it have to go through someone else we have to mention here?

alexbiehl commented:
I will do it. Is this ready then? All of @gbaz remarks fixed?

NunoAlexandre commented:
To me, it is ready. I am not for some of the @gbaz changes he made on top of mine in hackage, namely the reduced headings spacing, two-column view of package pages, etc. Those changes I am not making myself for 1) lack of time and 2) because they are not part of my approach. You guys are free to do what you find best on top of my work here, of course :)

gbaz commented on haddock-api/resources/html/Ocean.std-theme/ocean.css:
```diff
@@ -73,7 +73,7 @@ body.js-enabled .hide-when-js-enabled {
 @media only screen and (min-width: 1280px) {
   #content {
-    width: 70vw;
+    width: 63vw;
```
did you mean to revert 70 back to 63 here?

NunoAlexandre replied:
Yes. I tried 70vw but it left almost no space for the left column with the synopsis and table of contents, and the lines became too long to read. I am now on a 1280px screen, therefore the `#content` has 806px, which seems pretty fine to me. I guess most people will view these pages using a 1920px screen in full or split mode. When split, the viewport will have less than 1280px and it will make more use of the screen, meaning that in most cases the use of space is pretty good.

gbaz replied:
ah, didn't realize the changes made to the contents/synopsis to move them to the right. Yes, this is nice and fine.

gbaz commented:
One more issue I found -- in firefox we get bad behavior in the header when it is supposed to be one line. e.g.

![image](https://user-images.githubusercontent.com/1125849/37668825-be9bed66-2c3b-11e8-81de-b6da3ecfe681.png)

This was solved on hackage by reversing the order of the caption and ul like so: https://github.com/haskell/hackage-server/commit/7c94de40910def6bf36cb287c0d6721c0c84c7b4#diff-d2a08baffb29792b990326b60f1a891eL3

NunoAlexandre commented:
Thanks, @gbaz, good catch! I just fixed it 👍

ericnething commented:
@NunoAlexandre Having been quite happy with Hackage for many years now, I am extremely upset about the changes to the layout and styles.

What immediately stands out to me is the following:

* __Legibility has been reduced__ by using an extremely thin and small font combined with a __decrease__ in the contrast between the font color and the background color
* The most useful information is now crammed into a tiny column off to the far right
* The new color for links is very hard to see, and distinguishing between visited and unvisited links is nearly impossible
* Fonts are being loaded from Google Fonts (Why?) which breaks offline docs

__The changes are a net negative__ over the original haddock styles, which were perfectly functional. Your new styles have __decreased__ the usability of the documentation. I have to wonder if you took into account how people actually use these at all. The first thing I look at when I go to a package page is all of that information you have shoved off to the right in a tiny column.

![bad-haddocks](https://user-images.githubusercontent.com/8010421/37688274-9858c8c4-2c75-11e8-8719-1616190b6e41.png)

__The fonts on my screen are nearly unreadable__ because they are so thin, and the decreased contrast makes the text blend into the background. It hurts my eyes and forces me to squint. This is what it looks like if I decrease my browser window's width to be quite small. Notice the borked header.

![screenshot-2018-3-20 lens lenses folds and traversals 3](https://user-images.githubusercontent.com/8010421/37688900-9d1f0938-2c78-11e8-89ea-2cb96b14b1ab.png)

 Here is what it _could_ look like with some very minor changes if it were designed to be readable (sans borked header and ignoring the hyperlink color issue).

![screenshot-2018-3-20 lens lenses folds and traversals 2](https://user-images.githubusercontent.com/8010421/37688904-a2c01b20-2c78-11e8-912e-64bd8e2dc2a2.png)

This is using the following styles:

~~~css
body {
    font-weight: 400;
    font-size: 1.18rem;
    line-height: 1.4;
    font-family: sans-serif;
    color: #000000;
    background: #ffffff;
}
~~~

The most important things to take away from this are:

- __If I can't read it, then it is useless__
- __The most important information needs to come first__

For an example of how to design documentation correctly, look to [the Racket docs](https://docs.racket-lang.org/teachpack/2htdpuniverse.html). The Racket documentation was designed by Matthew Butterick, who is a designer and typographer who knows his shit. He wrote a book on typography called [Practical Typography](https://practicaltypography.com/). Regarding the Racket documentation, the fonts are perfectly chosen, and the most readable I have ever seen online. The text width is just right, the line spacing, the contrast --- everything! Even the presentation of the code examples is well thought-out. The more we can emulate the Racket documentation, the better. The current state of this redesign is simply unacceptable and must not stand. It is a step backwards.

gbaz commented:
@ericnething thank you for the useful concerns regarding readability. However, I find your tone rude and unhelpful, and in general I think you would be better received if you acted in a friendly spirit. A lot of work and discussion has gone into the redesign. It is of course not perfect yet -- getting these things right is work. We're all volunteers here, trying to improve things, and it is good to conduct discussions bearing that in mind.

First off, you should separate concerns about hackage, which belong on the hackage tracker, with concerns about general shared design, which can go on either tracker. Don't blame nuno for the two-col thing -- that's all on me, and was conducted on the basis of it being well received in recent days, and on the basis of many people disagreeing with you about which information is most useful (and on the basis of looking at other package repos and how they present information, such as npm and crates.io). But please feel free to open an issue on hackage about the two-col design, if you would like.

Moving on to the general design things, I think you raise a few important points.

Most important is offline docs. You're right that we can't rely on webfonts -- haddock documentation should be self-contained, so even if we have webfonts by default we need to ensure we have good fallbacks for the offline case. I agree that some good offline story, one way or another, is essential.

Second is link colors. I agree it is too hard to distinguish visited and unvisited (and the current link colors are my fault too, since I tweaked nuno's). I'm not sure of the right fix here? Either making visited a different color (rather than a similar one but desaturated), or lightning up further the unvisited links? Any proposal for a good scheme consistent with the overall palette would be welcomed.

Finally there's font-weight. I don't have strong opinion here. I think what you have is far too heavy and clunky. However, a bit more heaviness might help. One thing that strikes me about the racket docs is that they use a serif font -- which contravenes what people consider "modern looking" design, but really _does_ improve readability for large blocks of text. It is certainly worth considering!

Also: the header isn't borked. It is designed to flow on two lines, by default, as the browser resizes. We would like to swap the order of the two lines (and I think nuno figured out how!) but that's the extent of it.

ericnething commented:
@gbaz I'm sorry about my tone. I know that I didn't approach this in the most friendly way. Perhaps I am too invested in Haskell, but I care about this a lot and was shocked to see such a sudden change (I just found out about this tonight).

The Racket docs use several custom versions of the Chartered font. For open source fonts, that is the best you are going to find. The highest quality fonts (in general) are proprietary. Instead of using Google Fonts (and in particular Open Sans, which is massive in filesize), I suggest either hosting the font on the hackage server (and setting the fallback to comparable system fonts for each of Linux, Mac, and Windows instead of a simple `sans-serif`), or just going with the system fonts on their own.

Serif fonts always improve legibility, especially at smaller sizes. The only reason sans-serif fonts are so widely used on computers is because displays used to have resolutions too low to properly display the serifs. That is no longer true today and hasn't been for a long time.

One other thing I did not mention earlier is syntax highlighting for code samples (and code samples in general). We could use skylighting to do this (what pandoc uses), but the parser kind of sucks. I personally use pygments to process all of my haskell because the parser is the most correct. It requires installing python and the pygments library, but it is worth it. There might be a way to fix the buggy parser in skylighting, and that will eventually need to happen, but pygments is just easier.

I have been working on the styles and syntax highlighting for my book on Haskell called [Haskell Roguelike](http://haskellroguelike.com/chapter-4.html). I'm modeling it after the Racket docs, although it still doesn't come close. I'm using the Trac theme for syntax highlighting and pygments for parsing the code. Please ignore the shitty webfonts (they are temporary) and the wonky hyperlink colors.

gbaz commented:
In my opinion, merging sooner is better, but we should just feel pretty confident before a full release.

gbaz commented:
(Once it is merged I can try to use the stylesheet, or an adaptation of it that works with the old html at least, on hackage [through rewriting the existing haddocks on the fly], and that will help to get more iterative community feedback)

alexbiehl commented:
Recent haddocks write a meta.json for meta data. Do you think it would make sense to "version" the html so that Hackage could augment with a suitable stylesheet in cases like this? This would somewhat disentangle us here.

gbaz commented:
I think it would be fine if the meta.json just provided a haddock version number? Pegging ourselves to that is probably easier than trying to track potential moving parts that change at different places independently...

alexbiehl commented:
Then you are good go, haddock version is included already and Hackage knows
how to read that from meta.json!

gbaz <notifications@github.com> schrieb am Mi., 21. März 2018, 06:44:

> I think it would be fine if the meta.json just provided a haddock version
> number? Pegging ourselves to that is probably easier than trying to track
> potential moving parts that change at different places independently...
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/haskell/haddock/pull/721#issuecomment-374838943>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AByiiU9h8WcqsFNxnJxpzcSpxE4mKIzbks5tgejHgaJpZM4RM5HQ>
> .
>

lthms commented:
> By the way, why do you suggest hosting the font on hackage? Isn't google's service (which has all their CDNs) more efficient?

Google Fonts have raised concerned about privacy for years now. I don’t have a strong opinion about that, but I don't think regular users will notice a difference rather the fonts come from Google or Hackage itself; privacy aware will definitely enjoy having hackage hosting its fonts.

Also, it would be possible to cache fonts so they are download only once, no?

That being said, thanks a lot for all the efforts you made on redesigning hackage and haddock.

alexbiehl commented:
Please see https://github.com/haskell/haddock/pull/782 for the rebased patch set.

Profpatsch commented:
Just a heads-up, on 125% zoom (which I quite like for Haddocks), the two-column view currently leaves the screen to the right, while there is some space to the left:

![screenshot](https://user-images.githubusercontent.com/3153638/37708754-1ef5dc18-2d08-11e8-9f99-c3e399120962.png)

Otherwise very good job, I like the new design.

ericnething commented:
@gbaz I cannot agree with you on the font. The choice of font is simply wrong. It is illegible to my eyes and on my screen. Open Sans is an incredibly thin font. Even changing it to something like Lato makes it infinitely more readable. If I have trouble reading the documentation, it may as well not exist, because at that point it is useless. __The entire purpose of documentation is to be able to read it and obtain information.__

This is near impossible for me to read and hurts my eyes.

![screenshot-2018-3-21 introduction hackage 1](https://user-images.githubusercontent.com/8010421/37730527-5458e612-2d16-11e8-9439-b68ac640bb8a.png)

Here is what it looks like with 18px Lato (which comes standard on Mac OS, and there are equivalents on Linux and Windows, so there is no need to load a web font).

![screenshot-2018-3-21 introduction hackage](https://user-images.githubusercontent.com/8010421/37730443-1dbbe17c-2d16-11e8-93ee-9f439f50c2a0.png)

On how many eyes and how many displays have you tested your assumptions? I won't back down on the font. It is unacceptable.

Google Fonts should not be used. A third-party proprietary service that tracks you, harvesting and selling data on your activities, has no business being part of open source documentation for an open source programming language. I do not want them involved in any way with my documentation.

hvr commented:
@ericnething I'd like to try your 18px Lato font suggestion locally; how do I need to modify my CSS file in order to do that?

ericnething commented:
@hvr I was mistaken that Lato comes standard with Mac OS, but you can download it [here](http://www.latofonts.com/download/Lato2OFL.zip). When I have time I can put together a confirmed list of good system fonts and their equivalents for each major OS.

Right-click on any paragraph on the page and inspect it with your web console. It should be under `body` in the css.

~~~
body {
  font-family: 400 18px/1.5 Lato;
}
~~~

gbaz commented:
Bear in mind that Open Sans is indeed somewhat light, but the bulk of the difference between the screenshots is that the font-size is just set to be larger in the second one. Honestly I have a hard-time telling the difference between hevetica-alikes and don't feel too strongly about this in any particular way. Even just going with `verdana, sans-serif` is not too bad. It is also fine for haskell.org to host whatever web font may play a role, or not, in things.

Profpatsch commented:
@ericnething I find your conduct to be very condescending.

As for the font, Open Sans is kind of a standard, pretty legible font. Github starts their `font-family` with `-apple-system, BlinkMacSystemFont, `, which should probably be copied for Hackage. The `sans-serif` fallback already exists, so there’s that.

If something is “incredibly thin”, “hurts” a user’s eyes and there’s an “infinitely better” alternative one can download, then there’s always the possibility to use a custom user script.

NunoAlexandre commented:
@ericnething I do think you are exaggerating, not making less of your complaint. The contrast between the text color and the background color is in the highest category there is, and Open Sans is not _that_ light. I do consider your complaint because I once came across [this docs](https://dev.liferay.com/discover/portal/-/knowledge_base/6-2/what-is-liferay) and I experienced exactly the pain you describe, but while this is completely crazy, what is now live on hackage is rather acceptable and high contrast.

About alternative fonts, I'd not go for Lato. It's too "stylish" for such docs, in my opinion. I think that either [Source Sans Pro](https://fonts.google.com/specimen/Source+Sans+Pro) or [Helvetica (Nueue)] would be a better choice. They are slightly more thick than Open Sans and are also great to read such type of content like this.

Bear in mind that it's impossible to satisfy everyone. My eyes hurt with the previous hackage theme, while you apparently were happy with it. The majority of the feedback we got liked the new theme. A good number of comments about the text contrast was made and addressed properly, apart from the alternative fonts I made I don't see it is reasonable to make the font bold to make a very small minority happy.

Hrothen commented:
Load times are noticeably worse for me with the change. Google Fonts seems like the most likely culprit.

I also agree that the font is too light.

The two column format is very hard to read, at the minimum the metadata column should have a visible border, and the categories should probably be left justified. A better solution would probably be to actually put it on the side instead of sharing space with the description with empty space on both sides.

Actually looking through the metadata column is annoying now. I spent a good ten seconds trying to find the changelog for a package today.

Why not stick with the old format but make the metadata portion collapsed by default, like with typeclasses?

I also really dislike the new color scheme.

> The majority of the feedback we got liked the new theme

I just want to point out that the chances that even 10% of people who use Hackage were aware of this change coming are tiny.

gbaz commented:
@Hrothen you are on the haddock issue tracker. If you have comments on the two-col layout, I suggest you comment on a two col layout ticket on the hackage tracker. There are a number of tickets in-flight there. e.g. https://github.com/haskell/hackage-server/issues/721

Vis a vis load times, it can't be google fonts, or at least, it can't _continue_ to be, since they are cached for up to a year. So unless you have turned off all caching in your browser, they should cause negligible load.

Hrothen commented:
> There are a number of tickets in-flight there.
Man, this is way more awkward than one big thread.
So, what. I put my two column layout bits in that ticket and then open my own ticket to say "I don't like the colors"?

gbaz commented:
@NunoAlexandre  -- if you want to move to one of the other fonts you suggested, perhaps that might be worth it, since at least this font seems too light for _some_ people? I think the plan of attack -- pushing the tweaks against hackage-server first to get iterative "live" feedback -- and then porting things over here when they are "settled" is reasonable. (btw: commits here should be made against the rebased PR here, I think: https://github.com/haskell/haddock/pull/782).

gbaz commented:
@Hrothen I don't think you need to open a ticket to say "I don't like the colors." That would not be a very useful ticket. If you have concrete, actionable, changes, such as proposing a more visible set of colors for links, then that might be useful. Otherwise, you can consider your general dislike of colors sufficiently noted.

Hrothen commented:
If we're only supposed to open tickets if we have proposed changes, where are you collecting feedback?

gbaz commented:
Your feedback was already given that "I don't like the colors" right here. I don't think that adding it anywhere else will help things.

Cmdv commented:
@NunoAlexandre I think this is a vast improvement, I know it lacks the 1990's look some are so attached to but my eye wept at the new beauty 😍

ericnething commented:
I am continually frustrated by the most important information being shoved off into a tiny column on the right. Today I wanted to check on what updates have been made to Hakyll, and I had to scroll all the way to the bottom of the page to see when it was last updated, and what the latest version is. The two-column layout is absolutely terrible. Please fix it.

gbaz commented:
@ericnething Again, for the last time. This is the Haddock repo, not the hackage-server repo. Your comment pertains to two-col layout on hackage server, and is completely irrelevant to this PR. If you make any further comments here off topic, I will delete them. Please comment in the appropriate places, and please do so constructively.

ericnething commented:
@gbaz I am not the only one who is confused about where to provide feedback (see @Hrothen above) since you are spreading out these changes across both repos.

ntc2 commented:
This PR was closed accidentally, see [this comment](https://github.com/haskell/haddock/pull/814#issuecomment-387810373) for how to reopen.
